src/NLog/Filters/WhenRepeatedFilter.cs

```diff
@@ -0,0 +1,231 @@
+        {
+            MruCount = 50;
+            TimeoutSeconds = 10;
+            MaxLength = 1000;
+            OptimizeBufferReuse = true;
+        }
+
+        protected override FilterResult Check(LogEventInfo logEvent)
+        {
+            lock (_mruFilter)
+            {
+                using (var targetBuilder = OptimizeBufferReuse ? ReusableLayoutBuilder.Allocate() : ReusableLayoutBuilder.None)
+                {
+                    string value = string.Empty;
+                    int hashCode = 0;
+                    if (OptimizeBufferReuse)
+                    {
+                        Layout.RenderAppendBuilder(logEvent, targetBuilder.Result, false);
+                        hashCode = Math.Min(targetBuilder.Result.Length, MaxLength);
+                    }
+                    else
+                    {
+                        value = Layout.Render(logEvent) ?? string.Empty;
+                        if (value.Length > MaxLength)
+                            value = value.Substring(0, MaxLength);
+                        hashCode = StringComparer.Ordinal.GetHashCode(value);
+                    }
+
+                    int optimalRemoveIndex = 0;
+
+                    for (int i = _mruFilter.Count - 1; i >= 0; i--)
+                    {
+                        var filterInfo = _mruFilter[i];
+
+                        if (filterInfo.FilterCount == 0 || (logEvent.TimeStamp - filterInfo.LastLogTime).TotalSeconds > TimeoutSeconds)
+                            optimalRemoveIndex = i;
+
+                        if (filterInfo.StringHashCode != hashCode)
+                            continue;
+
+                        if (OptimizeBufferReuse)
+                        {
+                            bool stringEquals = true;
+                            int length = Math.Min(targetBuilder.Result.Length, MaxLength);
+                            if (length != filterInfo.StringBuffer.Length)
+                                continue;
+
+                            for (int x = 0; x < length; ++x)
+                            {
+                                if (targetBuilder.Result[x] != filterInfo.StringBuffer[x])
+                                {
+                                    stringEquals = false;
+                                    break;
+                                }
+                            }
+                            if (!stringEquals)
+                                continue;
+                        }
+                        else
+                        {
+                            if (!string.Equals(filterInfo.StringValue, value, StringComparison.Ordinal))
+                                continue;
+                        }
+
+                        FilterResult filterResult = FilterResult.Neutral;
+                        if ((logEvent.TimeStamp - filterInfo.LastLogTime).TotalSeconds > TimeoutSeconds || logEvent.Level.Ordinal > filterInfo.LogLevel.Ordinal)
```

Comment:
this is a duplicate check? (`(logEvent.TimeStamp - filterInfo.LastLogTime).TotalSeconds > TimeoutSeconds`)